lib/platform/src/linux/linux_x11clipboard.c

```diff
@@ -0,0 +1,337 @@
+	Atom xa_string;
+	Atom xa_targets;
+	Atom xa_text;
+	// Observer thread
+	LCUI_Thread observer_thread;
+} clipboard;
+
+void ExecuteCallback(LCUI_BOOL timed_out)
+{
+	if (!timed_out) {
+		_DEBUG_MSG("Didn't time out canceling observer thread\n");
+		LCUIThread_Cancel(clipboard.observer_thread);
+	}
+	LCUI_ClipboardCallback callback = clipboard.callback;
+	LCUI_Clipboard clipboard_data;
+
+	if (!callback->running) {
+		_DEBUG_MSG("Tried to ExecuteCallback before copying started\n");
+		return;
+	}
+	if (timed_out) {
+		_DEBUG_MSG("Callback timed out\n");
+		clipboard_data = NULL;
+	} else {
+		size_t len = clipboard.text_len + 1;
+		wchar_t *wstr = malloc(sizeof(wchar_t) * len);
+		len = decode_utf8(wstr, clipboard.text, len);
+		wstr[len] = 0;
+		// Assign the data
+		clipboard_data = malloc(sizeof(LCUI_ClipboardRec));
+		clipboard_data->text = wstr;
+		clipboard_data->len = len;
+		clipboard_data->image = NULL;
+	}
+	callback->action(clipboard_data, callback->arg);
+	// Reset properties, so if something goes wrongly, we crash
+	// instead of having undefined behaviour
+	callback->arg = NULL;
+	callback->action = NULL;
+	callback->running = FALSE;
+}
+
+void ClipboardRequestTimeout(void *arg)
+{
+	LCUI_ClipboardCallback callback = clipboard.callback;
+	sleep_ms(CLIPBOARD_TIMEOUT);
+	if (callback->running) {
+		ExecuteCallback(TRUE);
+	}
+}
+
+void RequestClipboardContent(void)
+{
+	LCUI_X11AppDriver x11 = LCUI_GetAppData();
+	Display *display = x11->display;
+	Window window = x11->win_main;
+	// Atom CLIPBOARD = XInternAtom(display, "CLIPBOARD", FALSE);
+	Window clipboard_owner =
+	    XGetSelectionOwner(display, clipboard.xclipboard);
+	// No need to continue, we should have stored text already
+	// when copy was done
+	// This branch will only get executed once we implement copy event
+	if (clipboard_owner == window) {
+		_DEBUG_MSG("Clipboard owned by self\n");
+		ExecuteCallback(FALSE);
+		return;
+	}
+	if (clipboard_owner == None) {
+		_DEBUG_MSG("Clipboard owner not found\n");
+		return;
+	}
+	// @WhoAteDaCake
+	// TODO: needs error handling if we can't access the clipboard?
+	// TODO: some other implementations will try XA_STRING if no text was
+	// retrieved from UTF8_STRING, however, our implementation is not
+	// synchronous, so not sure how it would work, it needs further
+	// investigation
+	LCUIThread_Create(&clipboard.observer_thread, ClipboardRequestTimeout, NULL);
+	Atom XSEL_DATA = XInternAtom(display, "XSEL_DATA", FALSE);
+	XConvertSelection(display, clipboard.xclipboard, clipboard.xutf8_string,
+			  XSEL_DATA, window, CurrentTime);
+	_DEBUG_MSG("Clipboard content requested, expect SelectionNotify\n");
+}
+
+/**
+ * This function assumes that text pointer is only owned by clipboard
+ */
+void LCUI_LinuxX11SetClipboardText(wchar_t *text, size_t len)
+{
+	// X11 doesn't support wchar_t, so we need to send it regular char
+	char *raw_text = malloc((len + 1) * sizeof(char));
+	size_t raw_len = wcstombs(raw_text, text, len);
```

Comment:
I'm not sure `wchar_t` is a better clipboard text type than `char`, but after seeing this line of code I think we can use char `instead` of `wchar_t` so that the clipboard implementation doesn't have to do the string encoding and decoding, what do you think?

Reply:
If windows clipboard needs char as well, then I think we should convert at `textedit` level, if it supports wchar_t, then for the sake of consistence I'd keep it as it is